Autopilot/Src/IMU.cpp

```
@@ -0,0 +1,217 @@
+//Power management
+#define REG_PWR_MGMT_1 0x6B
+#define REG_PWR_MGMT_2 0x6C
+//For I2C management
+#define REG_I2C_IF 0x70
+//Device idenitifier
+#define REG_WHO_AM_I 0x75
+
+/*** REGISTER DEFINITION ENDS ***/
+ 
+bool ICM20602::isSPIBusDefined = false;
+bool ICM20602::dataIsNew = false;
+
+float measuredAccX, measuredAccY, measuredAccZ; //Acceleration readings
+float measuredGyroX, measuredGyroY, measuredGyroZ; //Gyroscope readings 
+float measuredTemp; //Temperature reading
+static SPIPort *spi_port;
+static SPISettings hspi_1;
+uint8_t *imuBuffer;
+StatusCode sensorSuccess;
+
+/*** MAIN BLOCK OF CODE BEGINS ***/
+ 
+void ICM20602::Init() {
+   if(!isSPIBusDefined) {
+      // Configures the SPI Settings
+      hspi_1.port = SPI_PORT1;
+      hspi_1.mode = SPI_MODE_0;
+      hspi_1.master = true;
+      hspi_1.frequency = 10;                                            
+      
+      // Creates SPI Object
+      spi_port = new SPIPort(hspi_1);
+      GPIOPin pin(GPIO_PORT_A, 1, GPIO_OUTPUT, GPIO_STATE_LOW, GPIO_RES_NONE, GPIO_FREQ_HIGH, 0);                             //GET VALUES LATER 
+      imuSlaveIdentifier = spi_port->add_slave(pin); //Adds slave pin to the spi_port class object
+
+      uint8_t *setup = 0;
+
+      //Full reset of chip 
+      *setup = 0x80;
+      *imuBuffer = REG_PWR_MGMT_1 | 0x00; //0x80 means we are reading; 0x00 means we are writing
+      uint8_t writeData[2] = {*imuBuffer, *setup}; // writeData[0] -> register address; writeData[1] -> write instructions
+      write_data(writeData);
+
+      //Verify chip is working properly
+      *imuBuffer = REG_WHO_AM_I | 0x80;
+      uint8_t dataFromIMUSensor[2] = {0x00, 0x00}; // dataFromSensor[0] -> dummy byte (contains SPI address); dataFromSensor[1] -> data byte
+      spi_port->set_slave(imuSlaveIdentifier);
+      sensorSuccess = spi_port->exchange_data(imuBuffer, dataFromIMUSensor, 1);
+
+      if(sensorSuccess != STATUS_CODE_OK) {
+         //Debug code here
+      }
```

Is there a specific way we would like error handling to be done?

---

getResult should return an error code if something goes wrong. dw about adding debug code in the constructor for the moment.